lib/model/model.go

```diff
@@ -730,11 +738,45 @@ func (m *Model) Request(deviceID protocol.DeviceID, folder, name string, offset
 		return protocol.ErrNoSuchFile
 	}
 
-	if flags != 0 {
-		// We don't currently support or expect any flags.
+	if flags != 0 && flags != protocol.FlagFromTemporary {
+		// We currently support only no flags, or FromTemporary flag.
 		return fmt.Errorf("protocol error: unknown flags 0x%x in Request message", flags)
 	}
 
+	if debug && deviceID != protocol.LocalDeviceID {
+		l.Debugf("%v REQ(in): %s: %q / %q o=%d s=%d f=%d", m, deviceID, folder, name, offset, len(buf), flags)
+	}
+
+	if flags == protocol.FlagFromTemporary {
+		err := m.requestFromTemporaryFile(deviceID, folder, name, offset, buf)
+		if err == nil || !os.IsNotExist(err) {
+			return err
+		}
+	}
+
+	return m.requestFromFile(deviceID, folder, name, offset, buf)
+}
+
+func (m *Model) requestFromTemporaryFile(deviceID protocol.DeviceID, folder, name string, offset int64, buf []byte) error {
+	name = defTempNamer.TempName(name)
```

Comment:
We could do something to validate the resulting path. We had that in Request() earlier but I think it's been removed. I.e. something the checks that the full path after filepath.Clean() and whatnot actually has the folder path as a prefix.